src/main/kotlin/io/libp2p/pubsub/gossip/GossipRouter.kt

```diff
@@ -65,27 +96,133 @@ open class GossipRouter : AbstractRouter() {
-            is Rpc.ControlPrune ->
-                mesh[controlMsg.topicID]?.remove(receivedFrom)
-            is Rpc.ControlIHave ->
-                iwant(receivedFrom, controlMsg.messageIDsList - seenMessages)
-            is Rpc.ControlIWant ->
-                controlMsg.messageIDsList
-                    .mapNotNull { mCache.messages[it] }
-                    .forEach { submitPublishMessage(receivedFrom, it) }
+            is Rpc.ControlGraft -> handleGraft(controlMsg, receivedFrom)
+            is Rpc.ControlPrune -> handlePrune(controlMsg, receivedFrom)
+            is Rpc.ControlIHave -> handleIHave(controlMsg, receivedFrom)
+            is Rpc.ControlIWant -> handleIWant(controlMsg, receivedFrom)
         }
     }
 
+    private fun handleGraft(msg: Rpc.ControlGraft, peer: PeerHandler) {
+        val topic = msg.topicID
+        // ignore GRAFT for unknown topics
+        val meshPeers = mesh[topic] ?: return
+        when {
+            isDirect(peer) ->
+                prune(peer, topic)
+            isBackOff(peer, topic) -> {
+                notifyRouterMisbehavior(peer, 1)
+                if (isBackOffFlood(peer, topic)) {
+                    notifyRouterMisbehavior(peer, 1)
+                }
+                prune(peer, topic)
+            }
+            score.score(peer) < 0 ->
+                prune(peer, topic)
+            meshPeers.size >= coreParams.DHigh && !peer.isOutbound() ->
+                prune(peer, topic)
+            peer !in meshPeers ->
+                graft(peer, topic)
+        }
+    }
+
+    private fun handlePrune(msg: Rpc.ControlPrune, peer: PeerHandler) {
+        val topic = msg.topicID
+        mesh[topic]?.remove(peer)?.also {
+            notifyPruned(peer, topic)
+        }
+        if (this.protocol == PubsubProtocol.Gossip_V_1_1) {
+            if (msg.hasBackoff()) {
+                setBackOff(peer, topic, msg.backoff.seconds.toMillis())
+            } else {
+                setBackOff(peer, topic)
+            }
+            if (score.score(peer) >= scoreParams.acceptPXThreshold) {
+                processPrunePeers(msg.peersList)
+            }
+        } else {
+            if (msg.hasBackoff() || msg.peersCount > 0) {
+                notifyRouterMisbehavior(peer, 1)
+            }
+        }
+    }
+
+    private fun handleIHave(msg: Rpc.ControlIHave, peer: PeerHandler) {
+        val peerScore = score.score(peer)
+        // we ignore IHAVE gossip from any peer whose score is below the gossip threshold
+        if (peerScore < score.params.gossipThreshold) return
+        if (peerIHave.computeIfAbsent(peer) { AtomicInteger() }.incrementAndGet() > params.maxIHaveMessages) {
+            return
+        }
+
+        val iWant = msg.messageIDsList - seenMessages.keys
+        val asked = iAsked.computeIfAbsent(peer) { AtomicInteger() }
+        val maxToAsk = max(0, min(iWant.size, params.maxIHaveLength - asked.get()))
```

Comment:
Is the number of maximum ids in an `IWANT` equal to the number of maximum ids in an `IHAVE` message?

Reply:
Yep, it's not the part of the spec, but here is the Go implementation doc on this param: 
https://github.com/libp2p/go-libp2p-pubsub/blob/master/gossipsub.go#L150-L155

Reply:
However I'm not sure about the logic of how this parameter is processed in Go. Seems slightly odd. Will revisit it again

Reply:
Hm, now looks ok to me. Add a test for this and fix a bug